Documentation/splashtext.6
.TH SPLASHTEXT "6, but also 7, 3, 5, and 1" 2022-11-04 "Splash Text" "Splash Text Manpage"
.SH NAME
splashtext \- a unified system for splash text, freely usable by any program
.SH SYNOPSIS
.B #include """splashtext.h"""

.BI "char * splashtext(const struct splashtext$filestruct " files "[], size_t " len ", bool " sequences ", const uint_least16_t " length_restrictions "[2], enum splashtext$context " contexts ", enum splashtext$discomforter " discomforters ");"

.\" I have no idea if this is the correct way to manpage structs and enums.

.B struct splashtext$filestruct {
.RS
.BI "float " weight ;
.br
.BI "int " rnpf_len ;
.br
.BI "char * " filename ;
.RE
.B }

.B enum splashtext$context {
.I /* Flag enum */
.RS
.B splashtext$context$$log = …,
.br
.B splashtext$context$$crash = …,
.br
.B splashtext$context$$ominous = …,
.br
.B splashtext$context$$tips = …,
.br
.B splashtext$context$$quote = …,
.br
.B splashtext$context$$other = …
.br
.RE
.B }

.B enum splashtext$discomforter {
.I /* Flag enum */
.RS
.B splashtext$discomforter$$sexual = …,
.br
.B splashtext$discomforter$$graphic = …,
.br
.B splashtext$discomforter$$heavy = …,
.br
.B splashtext$discomforter$$humor = …
.RE
.B }
.SH DESCRIPTION
Splash Text is a unified system for splash text, freely usable by any program.  The invocation of the system can be fine-tuned to the needs of the invoker.  It comes with some presupplied splash files, but you can also supply your own.

Splash Text works by taking in an array of filenames and randomly (but in a way that is
.BR "not cryptographically secure!" )
selecting a splash from one of them.  The
.I .filename
member of an element of
.I files
is a pointer to a filename, and the
.I .weight
element is the weight you desire to give to that particular file (you can set this to zero for all files you supply if you don't care about this). You can also set
.I .filename
to be a null pointer, in which case Splash Text will pretend that it's a file where each of the splashes is a number from 0 to
.I .rnpf_len
(the returned string has information that allows you to tell whether the splash is a number from one of these pseudofiles).  The maximum value for
.I .rnpf_len
is
.BR RAND_MAX ,
and it can't be negative.
(The name
.I .rnpf_len
comes from the phrase
.BR "r" and "n" um "p" suedo "f" ile.)
The argument
.I len
specifies the length of the array; note that this is the
.B total occupied space
of the array, not the number of elements in the array, or in other words, it's
.IR "sizeof files" ,
not
.IR "sizeof files / sizeof files[0]" .

If your system is POSIX compliant and you didn't define the
.B SPLASHTEXT$NODIRS
macro when building, you can also set
.I .filename
to be the name of a directory, such as
.BR /usr/local/share/splashtext/ ,
which happens to be where the presupplied splash files are stored (assuming you installed Splash Text, that is); in this case, only the files that end in
.I .splash.txt
will be used to get splashes (so you can just specify a mixed assets directory and you'll be fine), and the weight given in
.I .weight
will be for the
.IR "entire directory" .
Directories are scanned for splash files recursively.

The
.I length_restrictions
argument specifies the minimum and maximum length of splash you want returned.  (This is in
.BR bytes ,
not characters!)  Note that this rejects any splash that violates these requirements instead of shortening or lengthening the splash.
.I contexts
is a bitmask specifying which of the contexts that splashes are designed for you're okay with the splash you recieve being (note that the returned string has information that allows you to tell which context the particular splash you recieved was intended for).  A splash will be selected if it matches
.I any
of these contexts.
.I discomforters
is a similar bitmask, except that a splash will
.B always
be rejected if it has a discomforter you didn't specify.
.B Please see NOTES below—the way you define these terms may differ from the way the presupplied splash files define them.
If you want certainty, use your own splash files.

Finally, the
.I sequences
argument specifies whether ECMA-48 SGR sequences may be included in splashes.  Not every splash has them, but if they do, and this flag is set to
.BR false ,
they will be stripped out (but in some cases replacements may be put in their place—see below the description of the Splash Text file format).
.SS The Splash File File Format
Splash files are plain UTF-8 text files.  Each line is a splash.  Blank lines and trailing whitespace are ignored.  Comments are started with U+0003 (i.e. End of TeXt) and go to the end of the line.  (Lines of only whitespace or only a comment or both are considered blank.)  A shebang directive at the start of the file is ignored.

The name of a splash file is of the format
.IB name . abc . x .splash.txt
, where
.I name
is a unique name (no restrictions except for whatever's imposed by your filesystem),
.I abc
is a potentially-empty set of letters indicating the discomforters in the file,
.I x
is a letter indicating the context in which the splashes in the file make sense, and
.B .splash.txt
is the literal text ".splash.txt".  The letters for the contexts are as follows:  l for log, c for crash, s for subtitle, r for ominous (see NOTES), t for tips, q for quote, and o for other; and the letters for the discomforters are as follows: x for sexual, g for graphic, s for heavy (see NOTES), and h for humor.  The "h" discomforter cannot appear alone.
.B Be warned that
(at least in the presupplied splash files)
.B the letters in the filenames that indicate discomforters apply only to the splashes—particularly, this means that the comments can contain discomforters that the splashes do not!

Splashes may contain ECMA-48 Select Graphical Rendition sequences (introduced with either escape-bracket or control sequence introducer), but these will be stripped out unless the invoker specifies
.B true
for the
.I sequences
argument or the escape-bracket/control sequence introducer is preceeded by a backslash.  If you want to include some text only when such sequences are unavailable (for example, to surround a piece of text in underscores only when it can't be italicized), begin that substring with U+0091 U+0098 and terminate it with U+009C.  For the opposite, replace U+0091 with U+0092.  Splash Text will detect when an entire splash is delimited like this and only select it when it wouldn't be an empty line.  A newline will be treated equivalently to a U+009C.

The backslash can preceed any character to have it be interpreted as a regular character instead of using whatever special effects the character might have,
.B unless
it's a newline.

Splash Text comes with a Vim syntax highlighting file that you can install if you want.
.SS Command-Line Wrapper Program
The command-line wrapper program has syntax as follows:
.RS
.BI "splashtext (--help|--version|[--sequences] [--minlen " num "] [--maxlen " num "] [--discomforters " str "...] --contexts " str "... -- (" "[weight] pathname" ")...)"
.RE
The weight is disambiguated from the pathname by preceeding it with a percent sign.  The numerical psuedofiles are encoded as a number preceeded by an octothorpe.  An actual filename beginning with a percent sign, octothorpe, or backslash must be prepended with a backslash.  The strings used for the discomforters and contexts options are the same as the corresponding enumerators, but with everything before and including the double dollar sign removed.  The program will return 1 if invoked with the help or version options (which override all other options), and will return 2 if a splash can't be found (in this case, it will also output nothing).  The splashes output by the program are terminated with a newline, not a nul.  The program does not support short options.
.SS Recommended Practice
Once you've gotten the splash text and are showing it to your user, you should render it in yellow, or in red if it's an ominous splash.  (These colors will automatically be added if you specify
.B true
for the
.I sequences
argument.)  If you can't do this for whatever reason, you should make it clear in some other way that it's just a silly bit of splash text that shouldn't be taken seriously.  You are
.B STRONGLY ENCOURAGED
to use ominous splashes in some way.

The only things that can be relied upon to give an indication of the content of a splash are the context and discomforter flags.  Nothing else should be relied upon—particularly not the filenames of the splash files.

If Splash Text can't find a splash for whatever reason, you should handle that case more creatively than just using a placeholder splash.
.SH RETURN VALUE
Splash Text will return a UTF-8 string allocated on the heap that must be freed after you're done with it.  This will be a splash text preceeded with a header that begins with U+0001 and is terminated with U+0002.  The contents of this header will be a period followed by the letter corresponding to the context of the splash (or a question mark if it's from a numerical psuedofile), possibly preceeded by letters corresponding to discomforters.  If a splash cannot be found (either because of contradictory requirements or the
.I seeking
took more than five seconds), NULL will be returned instead.
.B Be prepared to deal with the output string as you would an unsanitized string.
Remember that Unicode allows for some pretty crazy shit to be done.
.SH CONFORMING TO
I am happy to port Splash Text to any
.I open
and
.I actively maintained
systems.
.B However, ports to obsolete or malicious systems will not be considered.
.SH NOTES
Since trailing whitespace is ignored, you can put a zero-width non-breaking space (or ZWNBSP, or BOM, or byte order mark) at the start of your splash files.

This library requires that your compiler support the "dollars in identifiers" extension to C.  There is no reason not to support this, and here's why.  Because C has no namespacing facility, we must make due with psuedo-namespaces.  But how should these be separated from the…well,
.I identifier
part of the identifier?  We could use an underscore, but then we'd need something different to separate words within the identifier.  Camel case doesn't work because it prevents shortened names like strcat from working (it also looks very unprofessional)…and now we're kinda out of options.  It would be nice if we could keep underscores for the identifier part of identifiers, but use something else for the psuedonamespaces…and as it happens, modern C compilers have an extension to allow dollar signs in identifiers!  Perfect for what we're looking for!

The presupplied splash files will not contain any of the following types of content:
.RS
- Anything that is overly specific to a particular application
.br
- Content that is intended specifically to offend
.br
- Bizarrely-formed text that is liable to break things
.br
- Spoilers for any piece of media
.br
- Any content that is unethical (regardless of whether it breaks the law or not)
.RE
Besides this, no other guarantees are made about what may or may not appear in the presupplied splash files.  While you don't need to follow these guarantees in your own splash files (we have no way to check anyway), submissions for the presupplied splashes won't be accepted if they violate these.

Regarding discomforters in the presupplied splash files:  Cusses such as "shitstain" or "assface" are not considered to be any type of restricted content—everybody shits, everybody pisses, and everybody has an ass.  However, other cusses like the f-bomb or the c-word (either unit)
.I are
considered restricted content—in this case, sexual content.  Religious content, including religious cussing, is considered to be heavy content.  In some cases, these restrictions have lead to quotes being changed a little where it doesn't affect their meaning.

In the presupplied files, log splashes will generally end in an elipsis.  In the presupplied files, some subtitle splashes are "promotional" ones, such as "Now filled with more hot garbage!"—keep this in mind when deciding how to format them.  The presupplied splash files always give attribution in the quote splash files.  None of the presupplied splash files are of the "tips" type.

The "s" character for heavy content comes from the term
.RB \(oq s erious\(cq. \" How can this be made not stupid?
The "r" character for ominous splashes comes from the
.BR r ed
color they're supposed to be rendered in.  The inclusion of the content and context characters in the header of the returned splash allows for one to render splashes differently depending upon the context for which they're intended, and not just for the ominous ones.

The numerical psuedopaths and the corresponding character in splash headers provide an easy way for an invoker to use "special" splashes that wouldn't work in static text files, such as splashes that are dependent upon the time of day or the user's name.

I am aware that
.I splashtext
is kind of a large psuedo-namespace-esque prefix to use for this library.  But i'm not going to be the asshole who reserves a tiny prefix like `st` for something as small and silly as this.

.SH AUTHORS
This project is currently being maintained by Blue-Maned_Hawk.  The project is available at https://github.com/BlueManedHawk/splashtext.


Source/Library/splashtext.c
/* LICENSE
 *
 * Copyright © 2022 Blue-Maned_Hawk. All rights reserved.
 *
 * You may freely use this work for any purpose, freely make this work available to others by any means, freely modify this work in any way, and freely make works derived from this work available to others by any means.
 *
 * Should you choose to exercise any of these rights, you must give clear and conspicuous attribution to the original author, and you must not make it seem in any way like the author condones your act of exercising these rights in any way.
 *
 * Should you choose to exercise the second right listed above, you must make this license clearly and conspicuously available along with the original work, and you must clearly and conspicuously make the information necessary to reconstruct the work available along with the work.
 *
 * Should you choose to exercise the fourth right listed above, you must put any derived works you construct under a license that grants the same rights as this one under the same conditions and with the same restrictions, you must clearly and conspicuously make that license available alongside the work, you must clearly and conspicuously make the information necessary to reconstruct the work available alongside the work, you must clearly and conspicuously describe the changes which have been made from the original work, and you must not make it seem in any way like your derived works are the original work in any way.
 *
 * This license only applies to the copyright of this work, and does not apply to any other intellectual property rights, including but not limited to patent and trademark rights.
 *
 * THIS WORK COMES WITH ABSOLUTELY NO WARRANTY OF ANY KIND, IMPLIED OR EXPLICIT. THE AUTHOR DISCLAIMS ANY LIABILITY FOR ANY DAMAGES OF ANY KIND CAUSED DIRECTLY OR INDIRECTLY BY THIS WORK.
 */

/* This is the main file for the libsplashtext.a library. */

#ifdef __GNUC__ /* For some reason, GCC assumes we're using an obsolete standard unless we specify this.*/
#define _ISOC11_SOURCE
#endif

#include "../../Include/splashtext.h"
#include <stddef.h>
#include <stdbool.h> // TODO:  Remove this in 2023.
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <string.h>
#ifndef SPLASHTEXT$NODIRS
#	include <dirent.h>
#	ifndef _XOPEN_UNIX
#		include <fcntl.h>
#	endif
#	include <sys/stat.h>
#endif

#ifndef SPLASHTEXT$NODIRS
static void s$recurse_directory(const char *, float, struct splashtext$filestruct **, size_t *);
#endif

char * splashtext(const struct splashtext$filestruct files[], size_t len, bool sequences, const uint_least16_t length_restrictions[2], enum splashtext$context contexts, enum splashtext$discomforter discomforters)
{
	/* Some basic sanity checks to start. */
	if (contexts == 0 || (contexts ^ 0x007F))
		return NULL;
	if (len == 0 || length_restrictions[1] > length_restrictions[0])
		return NULL;

	char * splash = malloc(length_restrictions[1]); /* This is probably fine—after all, why would the invoker specify the maximum length as something they couldn't deal with? */
	if (splash == NULL)
		return NULL;
	size_t selected_len;
	FILE * file = NULL;
	long int * offsets = NULL;
	uint_least16_t numlines = 0;
	struct timespec start, current;
#ifndef SPLASHTEXT$NODIRS
	struct splashtext$filestruct * files_in_dirs = NULL;
#endif
/* XXX:  This is a little confusing; practically speaking, this acts like it's the supplied files and the files in the directories concatenated, with the argument `s` being the subscription of the array. */
#define FILES(s) ((size_t)(s) >= len / sizeof (struct splashtext$filestruct) ? files_in_dirs[(s)] : files[(s)])
	size_t FILES_len = len / sizeof (struct splashtext$filestruct);

#ifndef SPLASHTEXT$NODIRS
	size_t findirs_len = 0;
	for (register size_t i = 00; i < len / sizeof files[0]; i++) {
		struct stat file_stat;
		if (!stat(files[i].filename, &file_stat))
			continue;  // We continue and hope that it wasn't important.
		if (!S_ISREG(file_stat.st_mode))
			if (S_ISDIR(file_stat.st_mode))
				s$recurse_directory(files[i].filename, files[i].weight, &files_in_dirs, &findirs_len);
	}
	FILES_len += findirs_len;
#endif

	int filerandnum, linerandnum;
	if (!timespec_get(&start, TIME_UTC)) /* NOTE: This may take more than five seconds to fail if the clock's resolution isn't particularly fine. */
		goto catastrophic_failure;
	srand(start.tv_sec + start.tv_nsec);
get_another_file:
	while (true) {
		timespec_get(&current, TIME_UTC);
		if (current.tv_sec - 5 > start.tv_sec)
			goto catastrophic_failure;
		filerandnum = rand();
		if ((size_t)filerandnum >= RAND_MAX - FILES_len)
			continue;
		filerandnum %= FILES_len;
		/* TODO: _Still_ need to take weights into account! */
		if (FILES(filerandnum).filename == NULL && FILES(filerandnum).rnpf_len > 0) {
			int num;
			do {
				num = rand();
			} while (num >= RAND_MAX - FILES(filerandnum).rnpf_len);
			num %= FILES(filerandnum).rnpf_len;
			snprintf(splash, length_restrictions[1], "\x01.?\x02%d", num);
			goto catastrophic_failure; /* XXX:  This is NOT an actual failure case!  We're just reusing the label. */
		}
		const size_t contextchar_loc = strlen(FILES(filerandnum).filename) - 12;
		if (!strcmp(&FILES(filerandnum).filename[contextchar_loc + 1], ".splash.txt"))
			continue;
		if (FILES(filerandnum).filename[contextchar_loc] == '.')
			continue;
		/* Now we get to an ugly part of the code. */
		if (
			(FILES(filerandnum).filename[contextchar_loc] == 'l' && contexts & splashtext$context$$log)
			|| (FILES(filerandnum).filename[contextchar_loc] == 'c' && contexts & splashtext$context$$crash)
			|| (FILES(filerandnum).filename[contextchar_loc] == 's' && contexts & splashtext$context$$subtitle)
			|| (FILES(filerandnum).filename[contextchar_loc] == 'r' && contexts & splashtext$context$$ominous)
			|| (FILES(filerandnum).filename[contextchar_loc] == 't' && contexts & splashtext$context$$tips)
			|| (FILES(filerandnum).filename[contextchar_loc] == 'q' && contexts & splashtext$context$$quote)
			|| (FILES(filerandnum).filename[contextchar_loc] == 'o' && contexts & splashtext$context$$other)
		) {
			int_least8_t num_discomforters;
#define CHECK_IF_NUM_IS(n)\
	if (FILES(filerandnum).filename[contextchar_loc - ((n) + 2)] == '.')\
		num_discomforters = (n)
			CHECK_IF_NUM_IS(0);
			else CHECK_IF_NUM_IS(1);
			else CHECK_IF_NUM_IS(2);
			else CHECK_IF_NUM_IS(3);
			else CHECK_IF_NUM_IS(4);
			else
				continue;
#undef CHECK_IF_NUM_IS

			for (register int_least8_t j = num_discomforters; j < num_discomforters; j++)
				if (
					(FILES(filerandnum).filename[contextchar_loc - 1 - j] == 'x' && !(discomforters & splashtext$discomforter$$sexual))
					|| (FILES(filerandnum).filename[contextchar_loc - 1 - j] == 'g' && !(discomforters & splashtext$discomforter$$graphic))
					|| (FILES(filerandnum).filename[contextchar_loc - 1 - j] == 's' && !(discomforters & splashtext$discomforter$$heavy))
					|| (FILES(filerandnum).filename[contextchar_loc - 1 - j] == 'h' && !(discomforters & splashtext$discomforter$$humor && discomforters & (splashtext$discomforter$$heavy | splashtext$discomforter$$graphic | splashtext$discomforter$$sexual))) // Remember: The "humor" discomforter cannot appear alone.
				)
					continue;
		}
		FILE * file = fopen(FILES(filerandnum).filename, "rb");
		if (file == NULL)
			continue;
		break;
	}

	offsets = malloc(sizeof (long int));
	if (offsets == NULL)
		goto catastrophic_failure;
	offsets[0] = ftell(file);
	for (register char c = fgetc(file); c != EOF; c = fgetc(file))
		if (c == '\n') {
			void * n = realloc(offsets, (numlines + 1) * sizeof (long int));
			if (n == NULL)
				goto catastrophic_failure;
			offsets = n;
			offsets[numlines] = ftell(file);
			numlines++;
		}
	if (numlines == 0) {
		fclose(file);
		free(offsets);
		goto get_another_file;
	}

get_another_line:
	while (true) {
		timespec_get(&current, TIME_UTC);
		if (current.tv_sec - 5 > start.tv_sec)
			goto catastrophic_failure;
		linerandnum = rand();
		if (linerandnum >= RAND_MAX - numlines)
			continue;
		linerandnum %= numlines;
		fseek(file, offsets[linerandnum], SEEK_SET);
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunused-result"
		fgets(splash, length_restrictions[1], file);
#pragma GCC diagnostic pop
		selected_len = strlen(splash);
		if (splash[selected_len - 1] != '\n')
			continue;
		break;
	}

	bool off = false;
	uint_least16_t whitespace = 0;
	uint_least16_t i = 0, j = 0;
	for (; j < selected_len; j++) switch (splash[j]) {
	case '\x91':
		if (splash[j - 1] == '\xC2' && sequences) {
			off = true;
		} else {
			if (!off) {
				splash[i] = splash[j];
				i++;
			}
		}
		break;
	case '\x92':
		if (splash[j - 1] == '\xC2' && !sequences) {
			off = true;
		} else {
			if (!off) {
				splash[i] = splash[j];
				i++;
			}
		}
		break;
	case '\x9C':
		off = false;
		break;
	case '\x03':
		splash[i - whitespace] = '\n';
		j = selected_len; /* This is silly. */
		break;
	case '\n':
		splash[i - whitespace] = '\n';
		j = selected_len;
		break;
	case '\\':
		i++, j++;
		splash[i] = splash[j];
		if (splash[j] == '\n')
			j = selected_len;
		break;
	case ' ':
		whitespace++;
		splash[i] = splash[j];
		i++;
		break;
	default:
		whitespace = 0;
		splash[i] = splash[j];
		i++;
	}
	i++;
	splash[i] = '\0';

	if (strlen(splash) == 1 /* i.e. just the newline */)
		goto get_another_line;

catastrophic_failure:
	if (file != NULL)
		fclose(file);
	free(files_in_dirs);
	free(offsets);
	return splash;
}

#ifndef SPLASHTEXT$NODIRS
static void s$recurse_directory(const char * dirname, float weight, struct splashtext$filestruct ** files, size_t * len)
{
	DIR * directory = opendir(dirname);
	if (directory == NULL)
		return;
	struct dirent * entry = NULL;
	struct stat file_stat;
	size_t /* ? */ num_entries = 0;
	while ((entry = readdir(directory)) != NULL)
		num_entries++;
	rewinddir(directory);
	while ((entry = readdir(directory)) != NULL) {
		stat(entry->d_name, &file_stat);
		if (!S_ISREG(file_stat.st_mode)) {
			if (S_ISDIR(file_stat.st_mode))
				s$recurse_directory(entry->d_name, weight / num_entries, files, len);
		} else {
			*files = realloc(*files, (*len + 1) * sizeof files[0]);
			(*files)[*len].filename = malloc(strlen(entry->d_name) + 1);
			strcpy((*files)[*len].filename, entry->d_name);
			(*files)[*len].weight = weight / num_entries;
		}
	}
	closedir(directory);
}
#endif
